packages/vite/src/node/server/middlewares/static.ts

```diff
@@ -123,16 +127,17 @@ export function serveRawFsMiddleware(
 
   // Keep the named function. The name is visible in debug logs via `DEBUG=connect:dispatcher ...`
   return function viteServeRawFsMiddleware(req, res, next) {
-    let url = decodeURIComponent(req.url!)
+    const url = new URL(req.url!, 'http://example.com')
     // In some cases (e.g. linked monorepos) files outside of root will
     // reference assets that are also out of served root. In such cases
     // the paths are rewritten to `/@fs/` prefixed paths and must be served by
     // searching based from fs root.
-    if (url.startsWith(FS_PREFIX)) {
+    if (url.pathname.startsWith(FS_PREFIX)) {
+      const pathname = getDecodedPathname(url)
       // restrict files outside of `fs.allow`
       if (
         !ensureServingAccess(
-          slash(path.resolve(fsPathFromId(url))),
+          slash(path.resolve(fsPathFromId(pathname))),
```

Comment:
A problem here is that vite allows direct access to files with `?` in the middle. See https://github.com/vitejs/vite/blob/d2f02a848bb2b4f8da0102f00988954d193de59d/packages/vite/src/node/plugins/resolve.ts#L401
So if we only check the pathname, we may not hit the correct entry in the safe list (`moduleGraph.safeModulesPath`). So looks like the full URL needs to be used for `ensureServingAccess`

Reply:
It seems that that issue existed from before. sirv will only return `/example` when the request URL is `/example?file`.
So to fix this, IIUC we will need another middleware which does `req.url = encodeURIComponent(decodeURI(req.url))` and calls sirv after that.